### PR TITLE
Added clonality, percent on target and percent duplicates

### PR DESCRIPTION
**Pull request by aidaanva**
Add all the functionality discussed in #6

### CODE REVIEW

**Comment by aidaanva** [comment]
@jfy133 you will need to show me how to set up the git actions for the testing

File: endorS.py [hunk 1]
```diff
@@ -11,111 +11,248 @@
...
+
+    #Calculate Percentage on target raw (aka endogenous raw)
     if totalReads == 0.0:
-        endogenousPre = 0.000000
-        print("WARNING: no reads in the fastq input, Endogenous DNA raw (%) set to 0.000000")
-    elif mappedPre == 0.0:
-        endogenousPre = 0.000000
-        print("WARNING: no mapped reads, Endogenous DNA raw (%) set to 0.000000")
+        endogenousRaw = 0.000000
+        print("WARNING: no reads in the fastq input, Percent on target raw (%) set to 0.000000")
```
Comment (jfy133) [inline endorS.py:70]:
```suggestion
        print("WARNING: no reads in the fastq input, percent on target raw (%) set to 0.000000")
```

File: endorS.py [hunk 2]
```diff
@@ -11,111 +11,248 @@
...
+        print("WARNING: no mapped reads, Percent on target raw (%) set to 0.000000")
     else:
-        endogenousPre = float("{0:.6f}".format(round((mappedPre / totalReads * 100), 6)))
+        endogenousRaw = float("{0:.6f}".format(round((mappedRaw / totalReads * 100), 6)))
 except:
-    print("Incorrect input, please provide at least a samtools flag stats as input\nRun:\npython endorS.py --help \nfor more information on how to run this script")
-    sys.exit()
-#Check if the samtools stats post-quality filtering have been provided:
+    if args.verbose:
+        print("No samtools flagstat --raw provided. \nWARNING: none of the Percent on target stats will be calculated")
```
Comment (jfy133) [inline endorS.py:78]:
```suggestion
        print("No samtools flagstat --raw provided. \nWARNING: none of the percent on target stats will be calculated")
```

File: endorS.py [hunk 3]
```diff
@@ -11,111 +11,248 @@
...
-        endogenousPost = 0.000000
-        print("WARNING: no mapped reads, Endogenous DNA modified (%) set to 0.000000")
-    else:
-        endogenousPost = float("{0:.6f}".format(round((mappedPost / totalReads * 100),6)))
+    mappedPost = float((re.findall(r'([0-9]+) \+ [0-9]+ mapped',contentsqF))[0])
+    #Calculation of Percent on target quality filtered (aka endogenous DNA post-quality filtering):
+    if args.raw is not None:
+        if totalReads == 0.0:
+            endogenousQF = 0.000000
+            print("WARNING: no reads in the fastq input, Percent on target raw (%) set to 0.000000")
```
Comment (jfy133) [inline endorS.py:89]:
```suggestion
            print("WARNING: no reads in the fastq input, percent on target raw (%) set to 0.000000")
```

File: endorS.py [hunk 4]
```diff
@@ -11,111 +11,248 @@
...
+        elif mappedPost == 0.0:
+            endogenousQF = 0.000000
+            print("WARNING: no mapped reads, Percent on target modified (%) set to 0.000000")
```
Comment (jfy133) [inline endorS.py:92]: shouldn't modified be `quality filtered` here for consistency?

also

```suggestion
            print("WARNING: no mapped reads, percent on target modified (%) set to 0.000000")
```

File: endorS.py [hunk 5]
```diff
@@ -11,111 +11,248 @@
...
+        else:
+            endogenousQF = float("{0:.6f}".format(round(( mappedPost / totalReads * 100),6)))
 except:
-    print("Only one samtools flagstat file provided")
-    #Set the number of reads post-quality filtering to 0 if samtools
-    #samtools flag stats not provided:
-    mappedPost = "NA"
+    if args.verbose:
+        print("No samtools flagstat --qualityfiltered provided. \nWARNING: Percent on target Modified stat will not be calculated")
```
Comment (jfy133) [inline endorS.py:97]:
```suggestion
        print("No samtools flagstat --qualityfiltered provided. \nWARNING: Percent on target modified stat will not be calculated")
```

same here

File: endorS.py [hunk 6]
```diff
@@ -11,111 +11,248 @@
...
+        contentsdeDup= deDup.read()
+    #Extract number of reads pre dedup
+    totalPreDedup = float((re.findall(r'^([0-9]+) \+ [0-9]+ in total',contentsdeDup))[0])
+    #Extract number of mapped reads pre-quality filtering:
+    mappedDedup = float((re.findall(r'([0-9]+) \+ [0-9]+ mapped ',contentsdeDup))[0])
+    #Check if --raw provided and calculate Percent on target postDedup
+    if args.raw is not None:
+        if totalReads == 0.0:
+            endogenousDeDup = 0.000000
+            print("WARNING: no reads in the fastq input, Percent on target postDeDup (%) set to 0.000000")
```
Comment (jfy133) [inline endorS.py:110]:
```suggestion
            print("WARNING: no reads in the fastq input, percent on target post-deduplication (%) set to 0.000000")
```

File: endorS.py [hunk 7]
```diff
@@ -11,111 +11,248 @@
...
+        elif mappedDedup == 0.0:
+            endogenousDeDup = 0.000000
+            print("WARNING: no mapped reads, Percent on target postDeDup (%) set to 0.000000")
```
Comment (jfy133) [inline endorS.py:113]:
```suggestion
            print("WARNING: no mapped reads, Percent on target post-deduplication (%) set to 0.000000")
```

File: endorS.py [hunk 8]
```diff
@@ -11,111 +11,248 @@
...
+        else:
+            endogenousDeDup = float("{0:.6f}".format(round((mappedDedup / totalReads * 100),6)))
+    #Calculate clonality (aka cluster factor)
+    try:
+        clonality=float("{0:.6f}".format(round(( totalPreDedup / mappedDedup),6)))
+    except ZeroDivisionError:
+        clonality = 0
+        print("WARNING: non mapped reads postDeDup and/or preDeDup, check your BAM file. Clonality set to 0.000000")
```
Comment (jfy133) [inline endorS.py:121]:
```suggestion
        print("WARNING: non mapped reads post- and/or pre-deduplication, check your BAM file. Clonality set to 0.000000")
```

File: endorS.py [hunk 9]
```diff
@@ -11,111 +11,248 @@
...
+    #Calculate Percentage of Duplicates
+    try:
+        percentDuplicates=float("{0:.6f}".format(round((((totalPreDedup - mappedDedup) / totalPreDedup) * 100),6)))
+    except ZeroDivisionError:
+        percentDuplicates = 0
+        print("WARNING: non mapped reads postDeDup and/or preDeDup, check your BAM file. Percent Duplicates set to 0.000000")
```
Comment (jfy133) [inline endorS.py:127]:
```suggestion
        print("WARNING: non mapped reads post- and/or pre-deduplication, check your BAM file. Percent duplicates set to 0.000000")
```

File: endorS.py [hunk 10]
```diff
@@ -11,111 +11,248 @@
...
+except:
+    if args.verbose:
+        print("No samtools flagstat --deduplicated provided. \nWARNING: Percent on target postDedup, Clonality and Percent of Duplicates stats will not be calculated!")
```
Comment (jfy133) [inline endorS.py:130]:
```suggestion
        print("No samtools flagstat --deduplicated provided. \nWARNING: Percent on target post-deduplication, clonality and percent of duplicates stats will not be calculated!")
```

File: endorS.py [hunk 11]
```diff
@@ -11,111 +11,248 @@
...
+                        "clonality": clonality,
+                        "percent_duplicates": percentDuplicates
+                        }
+                },
+            }
+            print("Percent on target raw (%):",endogenousRaw)
+            print("Percent on target quality filtered (%):",endogenousQF)
+            print("Percent on target postDedup (%):", endogenousDeDup)
+            print("Clonality:",clonality)
+            print("Percent Duplicates (%):", percentDuplicates)
```
Comment (jfy133) [inline endorS.py:175]:
```suggestion
            print("Percent duplicates (%):", percentDuplicates)
```

File: endorS.py [hunk 12]
```diff
@@ -11,111 +11,248 @@
...
+                        "percent_on_target_quality_filtered": endogenousQF,
+                        "percent_on_target_postdedup": endogenousDeDup,
+                        "clonality": clonality,
+                        "percent_duplicates": percentDuplicates
+                        }
+                },
+            }
+            print("Percent on target raw (%):",endogenousRaw)
+            print("Percent on target quality filtered (%):",endogenousQF)
+            print("Percent on target postDedup (%):", endogenousDeDup)
```
Comment (jfy133) [inline endorS.py:173]:
```suggestion
            print("Percent on target post-deduplication (%):", endogenousDeDup)
```

File: endorS.py [hunk 13]
```diff
@@ -11,111 +11,248 @@
...
+                    name : { 
+                        "percent_on_target": endogenousRaw,
+                        "percent_on_target_postdedup": endogenousDeDup,
+                        "clonality": clonality,
+                        "percent_duplicates": percentDuplicates
+                        }
+                },
+            }
+        print("Percent on target raw (%):",endogenousRaw)
+        print("Percent on target postDedup (%):", endogenousDeDup)
```
Comment (jfy133) [inline endorS.py:212]:
```suggestion
        print("Percent on target post-deduplication (%):", endogenousDeDup)
```

File: endorS.py [hunk 14]
```diff
@@ -11,111 +11,248 @@
...
+    # Raw + Dedup: Percent target Raw, Percent target postDedup, Clonality, Percent duplicates
+    elif args.deduplicated is not None:
+        jsonOutput={
+                "id": "endorSpy",
+                "plot_type": "generalstats",
+                "pconfig": {
+                    "percent_on_target": { "max": 100, "min": 0, "title": "Percent on target (%)", "format": '{:,.2f}'},
+                    "percent_on_target_postdedup": { "max": 100, "min": 0, "title": "Percent on target postdedup (%)", "format": '{:,.2f}'},
+                    "clonality": { "max": 100, "min": 0, "title": "Clonality", "format": '{:,.2f}'},
+                    "percent_duplicates": { "max": 100, "min": 0, "title": "Percent Duplicates (%)", "format": '{:,.2f}'}
```
Comment (jfy133) [inline endorS.py:200]:
```suggestion
                    "percent_duplicates": { "max": 100, "min": 0, "title": "Percent duplicates (%)", "format": '{:,.2f}'}
```

**Review by jfy133 — CHANGES_REQUESTED** [review]
Mostly formatting!

File: endorS.py [hunk 15]
```diff
@@ -11,111 +11,248 @@
...
-parser.add_argument('samtoolsfiles', metavar='<samplefile>.stats', type=str, nargs='+',
-                    help='output of samtools flagstat in a txt file (at least one required). If two files are supplied, the mapped reads of the second file is divided by the total reads in the first, since it assumes that the <samplefile.stats> are related to the same sample. Useful after BAM filtering')
+    usage='python %(prog)s [-h] [--version] -r [<samplesfile>.stats] -qF [<samplesfile>.stats] -dedup [<samplesfile>.stats]',
+    formatter_class=argparse.RawDescriptionHelpFormatter,
+    description=textwrap.dedent('''\
+        author:
+        Aida Andrades Valtueña (aida.andrades[at]gmail.com)
+        
+        description:
+        %(prog)s calculates Percent on target (aka Endogenous DNA) from samtools flagstat files and print to screen.
```
[the remark this answers is not in the document]
Comment (aidaanva) [reply endorS.py:21]: it will print the name of the program

**Comment by aidaanva** [comment]
@jfy133 please recheck the review, I think I did all the requested changes and updated the READ.me properly this time

File: README.md [hunk 16]
```diff
@@ -1,28 +1,39 @@
...
-endorS.py [-h] [--version] <samplesfile>.stats [<samplesfile>.stats]
+usage: python endorS.py [-h] [--version] -r [<samplesfile>.stats] -qF [<samplesfile>.stats] -dedup [<samplesfile>.stats]
 ```
 
 ## Author
 Aida Andrades Valtueña (aida.andrades[at]gmail.com)
 
 ## Description and options
 ```bash
+endorS.py calculates Percent on target (aka Endogenous DNA) from samtools flagstat files and print to screen.
```
Comment (jfy133) [inline README.md:14]:
```suggestion
endorS.py calculates percent on target (aka Endogenous DNA) from samtools flagstat files and print to screen.
```

File: README.md [hunk 17]
```diff
@@ -1,28 +1,39 @@
...
+The Percent on target reported will be different depending on the combination of samtools flagstat provided.
```
Comment (jfy133) [inline README.md:15]:
```suggestion
The percent on target reported will be different depending on the combination of samtools flagstat provided.
```

File: README.md [hunk 18]
```diff
@@ -1,28 +1,39 @@
...
+This program also calculates clonality (aka Cluster Factor) and percent duplicates when the flagstat file after duplicate removal is provided
```
Comment (jfy133) [inline README.md:16]:
```suggestion
This program also calculates clonality (aka 'cluster factor') and percent duplicates when the flagstat file after duplicate removal is provided
```

File: endorS.py [hunk 19]
```diff
@@ -1,8 +1,8 @@
 #!/usr/bin/env python3
 
-"""Script to calculate the endogenous DNA in a sample from samtools flag stats.
-It accepts can accept up to two files: pre-quality and post-quality filtering. We recommend
-to use both files but you can also use the pre-quality filtering.
+"""Script to calculate the Percent on target (aka Endogenous DNA), clonality, and percent of duplicates in a sample from samtools flag stats.
```
Comment (jfy133) [inline endorS.py:3]:
```suggestion
"""Script to calculate the percent on target (aka endogenous DNA), clonality, and percent of duplicates in a sample from samtools flag stats.
```

File: endorS.py (hunk 15 above)
Comment (jfy133) [inline endorS.py:21]:
```suggestion
        %(prog)s calculates percent on target (aka Endogenous DNA) from samtools flagstat files and print to screen.
```

File: endorS.py [hunk 20]
```diff
@@ -11,111 +11,248 @@
...
-                    help='output of samtools flagstat in a txt file (at least one required). If two files are supplied, the mapped reads of the second file is divided by the total reads in the first, since it assumes that the <samplefile.stats> are related to the same sample. Useful after BAM filtering')
+    usage='python %(prog)s [-h] [--version] -r [<samplesfile>.stats] -qF [<samplesfile>.stats] -dedup [<samplesfile>.stats]',
+    formatter_class=argparse.RawDescriptionHelpFormatter,
+    description=textwrap.dedent('''\
+        author:
+        Aida Andrades Valtueña (aida.andrades[at]gmail.com)
+        
+        description:
+        %(prog)s calculates Percent on target (aka Endogenous DNA) from samtools flagstat files and print to screen.
+        The Percent on target reported will be different depending on the combination of samtools flagstat provided.
```
Comment (jfy133) [inline endorS.py:22]:
```suggestion
        The percent on target reported will be different depending on the combination of samtools flagstat provided.
```

File: endorS.py [hunk 21]
```diff
@@ -11,111 +11,248 @@
...
+        This program also calculates clonality (aka Cluster Factor) and percent duplicates when the flagstat file after duplicate removal is provided
```
Comment (jfy133) [inline endorS.py:23]:
```suggestion
        This program also calculates clonality (aka cluster factor) and percent duplicates when the flagstat file after duplicate removal is provided
```

**Review by jfy133 — APPROVED** [review]
Final tweaks but LGTM otherwise

**Pull request merged**